### PR TITLE
Fix GetMap tile_buffer when labels are rotated

### PR DESCRIPTION
**Pull request by troopa81**
Follows #43949

It fixes an issue when labels are rotated and in the tile buffer extent area.

For instance, the cut label `3` and `49` should not be displayed in the following image 

WMS_GetMap_Tiled_Rotated_Labels_True_result

cc @karl-magnus

**Funded by Kristianstad city**

### CODE REVIEW

File: src/core/qgsmapsettings.cpp [hunk 1]
```diff
@@ -400,6 +400,28 @@ QPolygonF QgsMapSettings::visiblePolygon() const
...
+  const QgsMapToPixel &m2p = mapToPixel();
+
+  // Transform tilebuffer in pixel.
+  // Original tilebuffer is in pixel and transformed only according
+  // extent width (see QgsWmsRenderContext::mapTileBuffer)
+
+  const double mapUnitsPerPixel = mExtent.width() / sz.width();
+  const double buffer = mExtentBuffer / mapUnitsPerPixel;
+
+  poly << m2p.toMapCoordinates( -buffer,                              -buffer ).toQPointF();
```
Comment (nyalldawson) [inline src/core/qgsmapsettings.cpp:417]: can you fix the formatting here? (I realise it's matching QgsMapSettings::visiblePolygon(), but there's no need to propagate that)

File: src/core/qgsmapsettings.h [hunk 2]
```diff
@@ -467,6 +467,14 @@ class CORE_EXPORT QgsMapSettings : public QgsTemporalRangeObject
      * \since QGIS 2.8
      */
     QPolygonF visiblePolygon() const;
+
+    /**
+     * Returns the visible area as a polygon (may be rotated) with extent buffer included
+     * \see extentBuffer()
+     * \since QGIS 3.20
```
Comment (nyalldawson) [inline src/core/qgsmapsettings.h:474]:
```suggestion
     * \since QGIS 3.22
```
?

File: tests/src/python/test_qgsserver_wms_getmap.py [hunk 3]
```diff
@@ -1806,6 +1806,25 @@ def test_wms_getmap_tile_buffer(self):
...
+            "CRS": "EPSG:3857",
+            "WIDTH": "512",
+            "HEIGHT": "512",
+            "LAYERS": "wms_tile_buffer_labels_rotated",
+            "FORMAT": "image/png",
+            "TILED": "true"
+        }.items())])
+
+        r, h = self._result(self._execute_request(qs))
+        self._img_diff_error(r, h, "WMS_GetMap_Tiled_Rotated_Labels_True")
```
Comment (nyalldawson) [inline tests/src/python/test_qgsserver_wms_getmap.py:1826]: The text in the reference image is quite narrow, so the test will be rather flaky on different platforms. Can you use the reference test font (QGIS Vera sans) in bold and bump the font size here?

File: src/core/qgsmapsettings.cpp [hunk 4]
```diff
@@ -400,6 +400,28 @@ QPolygonF QgsMapSettings::visiblePolygon() const
   return poly;
 }
 
+QPolygonF QgsMapSettings::visiblePolygonWithBuffer() const
```
Comment (nyalldawson) [inline src/core/qgsmapsettings.cpp:403]: Would be helpful to also directly test this method as a unit test

**Comment by troopa81** [comment]
I take your comments into consideration @nyalldawson , thank you for reviewing

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**